django/core/files/storage/filesystem.py

```diff
@@ -64,6 +65,19 @@ def directory_permissions_mode(self):
             self._directory_permissions_mode, settings.FILE_UPLOAD_DIRECTORY_PERMISSIONS
         )
 
+    def get_available_name(self, name, max_length=None):
+        if self._allow_overwrite:
+            name = str(name).replace("\\", "/")
+            dir_name, file_name = os.path.split(name)
+            if ".." in pathlib.PurePath(dir_name).parts:
+                raise SuspiciousFileOperation(
+                    "Detected path traversal attempt in '%s'" % dir_name
+                )
+            validate_file_name(file_name)
+            return name
+        else:
+            return super().get_available_name(name, max_length)
```

Comment:
Can you explain why this change is required? I also believe it is untested.

Reply:
This change is needed so that the name is allowed even if there's already a file by that name. If I remove this change, the two `file_storage.tests.OverwritingStorageTests` tests fail.

Reply:
I think we need to add `file_storage.test_generate_filename.GenerateFilenameStorageTests.test_storage_dangerous_paths_dir_name` or create a very similar test for when to `FileSystemStorage` has `allow_overwrite=True`

Reply:
I updated the tests.

Reply:
I think instead of overwriting `get_available_name` here, we can overwrite `exists` to something like:
```python
    def exists(self, name):
        exists = os.path.lexists(self.path(name))
        if self._allow_overwrite:
            return False
        return exists
```
It requires some test updates (these need a think) but something like:
```diff
@@ -99,14 +99,14 @@ class FileStorageTests(SimpleTestCase):
         f = self.storage.open("storage_test", "w")
         f.write("storage contents")
         f.close()
-        self.assertTrue(self.storage.exists("storage_test"))
+        self.assertTrue(os.path.exists(os.path.join(self.temp_dir, "storage_test")))

         f = self.storage.open("storage_test", "r")
         self.assertEqual(f.read(), "storage contents")
         f.close()

         self.storage.delete("storage_test")
-        self.assertFalse(self.storage.exists("storage_test"))
+        self.assertFalse(os.path.exists(os.path.join(self.temp_dir, "storage_test")))

     def _test_file_time_getter(self, getter):
         # Check for correct behavior under both USE_TZ=True and USE_TZ=False.
@@ -275,10 +275,10 @@ class FileStorageTests(SimpleTestCase):
         """
         Saving a pathname should create intermediate directories as necessary.
         """
-        self.assertFalse(self.storage.exists("path/to"))
+        self.assertFalse(os.path.exists(os.path.join(self.temp_dir, "path/to")))
         self.storage.save("path/to/test.file", ContentFile("file saved with path"))

-        self.assertTrue(self.storage.exists("path/to"))
+        self.assertTrue(os.path.exists(os.path.join(self.temp_dir, "path/to")))
         with self.storage.open("path/to/test.file") as f:
             self.assertEqual(f.read(), b"file saved with path")

@@ -694,13 +694,11 @@ class OverwritingStorageTests(FileStorageTests):
         stored_name_1 = self.storage.save(name, f_1)
         try:
             self.assertEqual(stored_name_1, name)
-            self.assertTrue(self.storage.exists(name))
             self.assertTrue(os.path.exists(os.path.join(self.temp_dir, name)))
             with self.storage.open(name) as fp:
                 self.assertEqual(fp.read(), content_1)
             stored_name_2 = self.storage.save(name, f_2)
             self.assertEqual(stored_name_2, name)
-            self.assertTrue(self.storage.exists(name))
             self.assertTrue(os.path.exists(os.path.join(self.temp_dir, name)))
             with self.storage.open(name) as fp:
                 self.assertEqual(fp.read(), content_2)
@@ -722,13 +720,11 @@ class OverwritingStorageTests(FileStorageTests):
         stored_name_1 = self.storage.save(name, f_1)
         try:
             self.assertEqual(stored_name_1, name)
-            self.assertTrue(self.storage.exists(name))
             self.assertTrue(os.path.exists(os.path.join(self.temp_dir, name)))
             with self.storage.open(name) as fp:
                 self.assertEqual(fp.read(), content_1)
             stored_name_2 = self.storage.save(name, f_2)
             self.assertEqual(stored_name_2, name)
-            self.assertTrue(self.storage.exists(name))
             self.assertTrue(os.path.exists(os.path.join(self.temp_dir, name)))
             with self.storage.open(name) as fp:
```